### PR TITLE
Change visualization to file-based approach

### PR DESCRIPTION
**Pull request by Sibert-Aerts**
The meat of this commit is removing the HTTP server-based visualization approach, POCO and all dependencies thereon, and adding functionality for the simulator to simply produce json files holding all information useful to a visualization tool.

It also includes an unfinished webpage capable of visualizing these files, but don't pay much mind to that yet.

### CODE REVIEW

File: src/main/cpp/core/Atlas.h [hunk 1]
```diff
@@ -7,35 +7,58 @@
...
+	Atlas(const ClusterMap& cluster_map) : cluster_map(cluster_map) {}
 
 	/// Look up a cluster's position. If it isn't found, throw std::out_of_range.
-	const geo::GeoPosition& Lookup(const Key& key) const { return map.at(key); }
+	const geo::GeoPosition& LookupPosition(const ClusterKey& key) const { return cluster_map.at(key); }
 
-	/// Store a GeoPosition in the atlas.
-	auto Emplace(const Atlas::Key& key, const geo::GeoPosition& pos) -> decltype(Map().emplace(key, pos))
+	/// Look up a cluster's town. If it isn't found, throw std::out_of_range.
+	const Town LookupTown(const ClusterKey& key) const { return town_map.at(cluster_map.at(key)); }
```
Comment (jonathanvdc) [inline src/main/cpp/core/Atlas.h]: Did you mean `const Town&`? A return value of `const Town` is identical to a return value of `Town` in every way, except for the added visual noise.

File: src/main/cpp/http/prototype.h [hunk 2]
```diff
@@ -1,93 +0,0 @@
...
-using Poco::Util::ServerApplication;
-using Poco::Util::Application;
-using Poco::Util::Option;
-using Poco::Util::OptionSet;
-using Poco::Util::OptionCallback;
-using Poco::Util::HelpFormatter;
-
-namespace Stride {
-
-class HelloWorldRequestHandler : public HTTPRequestHandler
```
Comment (jonathanvdc) [inline src/main/cpp/http/prototype.h:45]: RIP HTTP server. :cry:

File: src/main/cpp/output/VisualizerData.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+public:
+	/// Register a new day from the given population's current state.
+	void AddDay(const Population& pop);
+
+	/// Get a reference to the vector of days.
+	const std::vector<std::map<std::size_t, unsigned int>>& GetDays() const;
+
+	/// Convert all stored data into a boost property tree, so that it can be converted to JSON.
+	const std::shared_ptr<boost::property_tree::ptree> ToPtree() const;
```
Comment (jonathanvdc) [inline src/main/cpp/output/VisualizerData.h]: Can you remove the `const` qualifier on the return value here? It doesn't do or mean anything.

File: src/main/cpp/output/VisualizerFile.h [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+{
+public:
+	/// Constructor: initialize.
+	VisualizerFile(const std::string& file = "stride_vis");
+
+	/// Destructor: close the file stream.
+	~VisualizerFile();
+
+	/// Print the given visualisation data.
+	void Print(Atlas::TownMap townMap, const VisualizerData& visualizer_data);
```
Comment (jonathanvdc) [inline src/main/cpp/output/VisualizerFile.h]: Is an `Atlas::TownMap` a big data structure? If so, can you pass it by reference?

File: src/main/cpp/core/Atlas.h [hunk 5]
```diff
@@ -7,35 +7,58 @@
...
 /**
- * A map from clusters, identified by their ID and type, to geopositions.
+ * A map storing geopositions for clusters, and town data for geopositions.
  */
 class Atlas
 {
 public:
-	using Key = std::pair<std::size_t, ClusterType>;
-	using Map = std::map<Key, geo::GeoPosition>;
+	struct Town
```
Comment (jonathanvdc) [inline src/main/cpp/core/Atlas.h]: Can you maybe document this `struct` a little bit?

File: src/main/cpp/core/Atlas.h [hunk 6]
```diff
@@ -7,35 +7,58 @@
...
+	{
+		std::string name;
+		unsigned int size;
+		static std::size_t newId;
```
Comment (jonathanvdc) [inline src/main/cpp/core/Atlas.h]: Does this have to be `public`? I figure that making this `private` would reduce the odds of someone secretly changing `newId`'s value and messing up all the IDs.

File: src/main/cpp/pop/Population.h [hunk 7]
```diff
@@ -153,13 +153,26 @@ class Population
...
-	auto AtlasEmplace(const Atlas::Key& key, const geo::GeoPosition& pos) -> decltype(atlas.Emplace(key, pos))
+	/// Store a Cluster's GeoPosition in the population's atlas.
+	auto AtlasEmplaceCluster(const Atlas::ClusterKey& key, const geo::GeoPosition& pos)
+	    -> decltype(atlas.EmplaceCluster(key, pos))
 	{
-		return atlas.Emplace(key, pos);
+		return atlas.EmplaceCluster(key, pos);
+	}
+
+	Atlas::Town Hometown(const Person& person) const{
```
Comment (jonathanvdc) [inline src/main/cpp/pop/Population.h]: If you change `atlas.LookupTown`'s return type to `const Atlas::Town&`, then you can have `Hometown` return a `const Atlas::Town&` as well (which should be slightly more efficient I guess).

Also, can you rename `Hometown` to something like `GetHometown` to make it easier to infer what this method does?

File: src/main/cpp/sim/run_stride.cpp [hunk 8]
```diff
@@ -92,9 +93,14 @@ void StrideSimulatorResult::AfterSimulatorStep(
 		cp->CloseFile();
 	}
 	*/
+	auto& pop = sim.GetPopulation();
 	run_clock.Stop();
 	auto infected_count = sim.GetPopulation()->get_infected_count();
 	cases.push_back(infected_count);
+
+	if(generate_vis_data && pop.has_atlas)
```
Comment (jonathanvdc) [inline src/main/cpp/sim/run_stride.cpp]: The spacing in this `if` statement looks a little off. Can you `clang-format` this file?

File: src/main/cpp/pop/Population.h [hunk 9]
```diff
@@ -153,13 +153,26 @@ class Population
 		}
 		return total;
 	}
+  
+	/// Get the atlas.
+	const Atlas& getAtlas() const {return atlas;}
```
Comment (jonathanvdc) [inline src/main/cpp/pop/Population.h]: The whitespace looks weird here. Can you `clang-format` this file?

**Review by jonathanvdc — APPROVED** [review]
This PR looks great! I've made some remarks, but they're mostly related to aesthetics and/or whether values are passed as references or not.

File: src/main/cpp/http/prototype.h (hunk 2 above)
Comment (Sibert-Aerts) [reply src/main/cpp/http/prototype.h:45]: :cry:

**Pull request merged**